onetimepad.1
.TH "Onetimepad" "1" "Version 0.1" "Sven Haardiek, Lars Kiesow" "Cryptographic tool"
.SH "NAME"
onetimepad - cryptographic toolkit for the command line

.SH "SYNOPSIS"
\fBonetimepad\fI [ \fI\-options\fR ] \fIcommand\fR [ \fIcommand_args\fR ]

.SH "DESCRIPTION"
\fBOnetimepad\fR encrypts and decrypts files by linking them XOR with a key files and should be used for secure data exchange with other people, because after exchange of the key file the encrypted communication should be save.

The keyfiles consist of random characters generated with the \fBOpenssl libary\fR (default) or the \fBGcrypt library\fR.

Before key files can be used they have to be imported by the program as a key file for encryption, decryption or both. Key files for encryption get the postfix .public. If the keyfile is imported to encrypt and decrypt, \fBonetimepad\fR will generate two files, because the used parts of the key files will be override by zeros but needed again for decryption (usually encryption and decryption are not running on the same pc).

After importing a key file it will be encrypted with the twofish algorithm from the \fBGcrypt library\fR with a 256bit key derivated from a passphrase. 

.SH "COMMANDS"
.SS "\fBgenerate\fR" [ \fIsize\fR ] [ \fIkeyfile\fR ]
Generate a key file of random characters with the size \fIsize\fR in kilobyte and the name \fIkeyfile\fR. On default the \fBOpenssl library\fR is used to generate random bytes.

.SS "\fBimport-enc\fR [ \fIkeyfile\fR ]
Import the key file \fIkeyfile\fR to use for encrytion. Therefore the file is renamed as \fIkeyfile.public\fR. The key file will be encrypted with twofish. 

.SS "\fBimport-dec\fR [ \fIkeyfile\fR ]
Import the key file \fIkeyfile\fR to use for decrytion. The key file will be encrypted with twofish. 

.SS "\fBencrypt\fR" [ \fIinputfile\fR ] [ \fIkeyfile\fR ]  [\fIoutputfile\fR ]
Encrypt the input file with the name \fIinputfile\fR using the key file \fIkeyfile\fR and saving the result in the output file with the name \fIoutputfile\fR. The keyfile must be imported for encryption (they get the postfix .public).

.SS "\fBdecrypt\fR" [ \fIinputfile\fR ] [ \fIkeyfile\fR ]  [\fIoutputfile\fR ]
Decrypt the input file with the name \fIinputfile\fR using the key file \fIkeyfile\fR and saving the result in the output file with the name \fIoutputfile\fR. The keyfile must be imported for decryption.

.SS "\fBkeyinfo\fR" [ \fIkeyfile\fR ]
Print the header information of the keyfile \fIkeyfile\fR to the standard output.

.SH "OPTIONS"
.IP "\fB\-h\fR"
Show a little help for usage.

.IP "\fB\-r\fR"
Toggle reseed of OpenSSL PRNG (default: reseed).

.IP "\fB\-k\fR"
Toggle if the key part used for decryption shall be kept to enable a second decryption. (default: overwrite).

.IP "\fB\-l\fR"
Change library to use for random number generation to \fBGcrypt library\fI.

.IP "\fB\-q\fR"
Use very strong pseudo random number generator.

.SH "RETURN VALUE"
Upon successful completion \fBonetimepad\fR return \fI0\fR. Otherwise is returned \fI1\fR.


onetimepad.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include "crypt.h"
#include "header.h"
#include "generate.h"
#include "import.h"
#include <getopt.h>
#include "twofish.h"

/* usage print a help */
void usage(const char* prgm) {
	printf( "\nUsage: %s [options] command [command options]\n\n"
			"-- Commands: ----\n"
			"\tgenerate        -- Generate a new keyfile\n"
			"\timport-enc      -- Import a keyfile for encryption\n"
			"\timport-dec      -- Import a keyfile for decryption\n"
			"\tencrypt         -- Encrypt a file\n"
			"\tdecrypt         -- Decrypt a file\n"
			"\tkeyinfo         -- Get information to a keyfile\n\n"
			"-- Options ----\n"
			"\t-h              -- Show this help\n"
			"\t-r              -- Toggle reseed of OpenSSL PRNG (default: reseed)\n"
			"\t-k              -- Toggle if the key part used for decryption shall be\n"
			"\t                   kept to enable a second decryption. (default: overwrite)\n\n"
			"\t-l              -- Change Library to use for encryption and random number generation to Gcrypt library\n"
			"\t-q              -- Use very strong pseudo random number generator\n"
			"For more informations read the man page.\n\n"
			,prgm);
}

/* Read in arguments an transfer them
 * to encrypt, decrypt, generate and import */
int main(int argc, char *argv[]) {

	/* Set default values 
	 * Default values:
	 * - Reseed
	 * - Do not keep keys 
	 * - use openssl 
	 * - use gcrypt random quality strong
	 * - Set import status decrypt and encrypt */
	config_t config = { 1, 0, USE_OPENSSL, GCRY_STRONG_RANDOM, STATUS_UNUSED_KEY };

	/* Read in options */
	int c;
	while ((c = getopt (argc, argv, "rklqs:h")) != -1) {
		switch (c) {
			case 'r': 
				config.reseed = !config.reseed;
				break;
			case 'k': 
				config.keep_key = !config.keep_key;
				break;
			case 'l':
				config.cryptlib = USE_LIBGCRYPT;
				break;
			case 'q':
				config.random_quality = GCRY_VERY_STRONG_RANDOM;
				break;
			case 'h':
				usage(*argv);
				return 0;
		}
   }

	/* Show help if there is no command  */
	if (argc - optind < 1) {
		usage(*argv);
		return 0;
	}

	/* Go through the commands */
	const char* cmd = argv[optind];

	/* Case: generate */
	if (!strcmp("generate", cmd)) {
		if (argc - optind == 3) {
			return generate( &config, atol(argv[optind+1]), argv[optind+2]);
		}
		printf("Command options: generate [size (kb)] [keyfile]\n");
		return 0;
	}

	/* Case: keyinfo */
	if (!strcmp("keyinfo", cmd)) {
		if (argc - optind == 2) {
			
			/* open keyfile */
			FILE * f = fopen(argv[optind+1],"r");
			if (!f) {
				fprintf(stderr, "Could not open keyfile %s", argv[optind+1]);
				return 1;
			}
			
			/* Create new header */
			header_t keyheader;

			/* Read header from keyfile */
			if (fread(&keyheader,1,sizeof(header_t),f) 
					!= sizeof(header_t) ) {
				fprintf(stderr, "Could not read from keyfile %s\n", argv[optind+1]);
				fclose(f);
				return 1;
			}

			printf("Key Status: %i\n", keyheader.status);
			printf("Key Position: %lu\n", keyheader.pos);
			printf("Key Id: %i\n", keyheader.id);
			printf("Key Size: %lu\n", keyheader.size);
		
			fclose(f);
			return 0;

		}
		printf("Command options: keyinfo [keyfile]\n");
		return 0;
	}

	/* Case: import key for encoding */
	if (!strcmp("import-enc", cmd)) {
		if (argc - optind == 2) {
			config.get_status = STATUS_ENC_KEY;
			return import( &config, argv[optind+1]);
		}
		printf("Command options: %s [keyfile]\n", cmd);
		return 0;
	}

	/* Case: import key for decoding */
	if (!strcmp("import-dec", cmd)) {
		if (argc - optind == 2) {
			config.get_status = STATUS_DEC_KEY;
			return import( &config, argv[optind+1]);
		}
		printf("Command options: %s [keyfile]\n", cmd);
		return 0;
	}

	/* Case: encrypt */
	if (!strcmp("encrypt", cmd)) {
		if (argc - optind == 4) {
			return encrypt(&config, argv[optind+1], argv[optind+2], argv[optind+3]);
		}
		printf("Command options: encrypt [infile] [keyfile] [outfile]\n");
		return 0;
	}

	/* Case: decrypt */
	if (!strcmp("decrypt", cmd)) {
		if (argc - optind == 4) {
			return decrypt(&config, argv[optind+1], argv[optind+2], argv[optind+3]);
		}
		printf("Command options: decrypt [infile] [keyfile] [outfile]\n");
		return 0;
	}

	/* Case: default */
	usage(*argv);
	return 0;
}
